File: internal/mmr/mmr.go
package mmr

import (
	"math/bits"
)

// MMR represents a Merkle Mountain Range structure.
// An MMR is an append-only authenticated data structure consisting of
// multiple perfect binary trees ("mountains") of decreasing size.
type MMR struct {
	store Store    // Backing store for nodes
	size  uint64   // Total number of nodes (not leaves)
	peaks []uint64 // Indices of current peak nodes
}

// New creates a new MMR with the given backing store.
func New(store Store) (*MMR, error) {
	mmr := &MMR{
		store: store,
		size:  0,
		peaks: make([]uint64, 0),
	}

	// Reconstruct state from store
	if err := mmr.restore(); err != nil {
		return nil, err
	}

	return mmr, nil
}

// restore rebuilds the MMR state from the backing store.
func (m *MMR) restore() error {
	size, err := m.store.Size()
	if err != nil {
		return err
	}
	m.size = size

	if size == 0 {
		m.peaks = make([]uint64, 0)
		return nil
	}

	m.peaks = findPeaks(size)
	return nil
}

// Append adds a new leaf to the MMR and returns its index.
// After inserting the leaf, it merges peaks of equal height.
func (m *MMR) Append(data []byte) (uint64, error) {
	// Create the new leaf node
	leafIndex := m.size
	leaf := NewLeafNode(leafIndex, data)

	if err := m.store.Append(leaf); err != nil {
		return 0, err
	}
	m.size++

	// Merge peaks of equal height
	if err := m.mergePeaks(); err != nil {
		return 0, err
	}

	// Recalculate peaks
	m.peaks = findPeaks(m.size)

	return leafIndex, nil
}

// mergePeaks merges adjacent peaks of equal height until all peaks have different heights.
func (m *MMR) mergePeaks() error {
	for {
		peaks := findPeaks(m.size)
		if len(peaks) < 2 {
			return nil
		}

		// Check if the last two peaks have the same height
		lastIdx := peaks[len(peaks)-1]
		prevIdx := peaks[len(peaks)-2]

		lastNode, err := m.store.Get(lastIdx)
		if err != nil {
			return err
		}

		prevNode, err := m.store.Get(prevIdx)
		if err != nil {
			return err
		}

		if lastNode.Height != prevNode.Height {
			return nil
		}

		// Merge: create a new internal node
		newNode := NewInternalNode(m.size, lastNode.Height+1, prevNode, lastNode)
		if err := m.store.Append(newNode); err != nil {
			return err
		}
		m.size++
	}
}

// GetPeaks returns the hash values of all current mountain peaks.
func (m *MMR) GetPeaks() ([][32]byte, error) {
	if m.size == 0 {
		return nil, nil
	}

	peaks := findPeaks(m.size)
	hashes := make([][32]byte, len(peaks))

	for i, idx := range peaks {
		node, err := m.store.Get(idx)
		if err != nil {
			return nil, err
		}
		hashes[i] = node.Hash
	}

	return hashes, nil
}

// GetRoot computes the "Witness Root" by hashing all peaks together.
// This single 32-byte value represents the entire document history.
func (m *MMR) GetRoot() ([32]byte, error) {
	if m.size == 0 {
		return [32]byte{}, ErrEmptyMMR
	}

	peaks, err := m.GetPeaks()
	if err != nil {
		return [32]byte{}, err
	}

	if len(peaks) == 1 {
		return peaks[0], nil
	}

	// Bag the peaks from right to left (standard MMR convention)
	root := peaks[len(peaks)-1]
	for i := len(peaks) - 2; i >= 0; i-- {
		root = HashInternal(peaks[i], root)
	}

	return root, nil
}

// Size returns the total number of nodes in the MMR.
func (m *MMR) Size() uint64 {
	return m.size
}

// LeafCount returns the number of leaf nodes in the MMR.
func (m *MMR) LeafCount() uint64 {
	return leafCountFromSize(m.size)
}

// Get retrieves a node by its index.
func (m *MMR) Get(index uint64) (*Node, error) {
	if index >= m.size {
		return nil, ErrIndexOutOfRange
	}
	return m.store.Get(index)
}

// GenerateProof creates an inclusion proof for the leaf at the given index.
// The proof contains the Merkle path from the leaf to the peaks, plus peak bagging.
func (m *MMR) GenerateProof(leafIndex uint64) (*InclusionProof, error) {
	if m.size == 0 {
		return nil, ErrEmptyMMR
	}

	if leafIndex >= m.size {
		return nil, ErrIndexOutOfRange
	}

	node, err := m.store.Get(leafIndex)
	if err != nil {
		return nil, err
	}

	// Only leaves can be proven
	if node.Height != 0 {
		return nil, ErrInvalidProof
	}

	// Find which mountain (peak) this leaf belongs to
	path, peakIndex, err := m.generateMerklePath(leafIndex)
	if err != nil {
		return nil, err
	}

	// Get peak hashes for bagging
	peaks, err := m.GetPeaks()
	if err != nil {
		return nil, err
	}

	peakIndices := findPeaks(m.size)
	var peakPosition int
	for i, idx := range peakIndices {
		if idx == peakIndex {
			peakPosition = i
			break
		}
	}

	root, err := m.GetRoot()
	if err != nil {
		return nil, err
	}

	return &InclusionProof{
		LeafIndex:    leafIndex,
		LeafHash:     node.Hash,
		MerklePath:   path,
		Peaks:        peaks,
		PeakPosition: peakPosition,
		MMRSize:      m.size,
		Root:         root,
	}, nil
}

// generateMerklePath computes the path from a leaf to its peak.
// MMR structure example (7 nodes):
//
//	    6
//	   / \
//	  2   5
//	 / \ / \
//	0  1 3  4
func (m *MMR) generateMerklePath(leafIndex uint64) ([]ProofElement, uint64, error) {
	var path []ProofElement
	pos := leafIndex

	// Get the node to verify it's a leaf
	node, err := m.store.Get(pos)
	if err != nil {
		return nil, 0, err
	}
	height := node.Height

	for {
		// Try to find sibling and parent by checking both possibilities
		siblingPos, parentPos, isRightChild, found := m.findFamily(pos, height)

		if !found {
			// No valid sibling/parent found, we're at a peak
			return path, pos, nil
		}

		// Get sibling hash
		siblingNode, err := m.store.Get(siblingPos)
		if err != nil {
			return path, pos, nil
		}

		// IsLeft in proof means sibling is on the left
		path = append(path, ProofElement{
			Hash:   siblingNode.Hash,
			IsLeft: isRightChild,
		})

		// Move to parent
		pos = parentPos
		height++
	}
}

// findFamily determines sibling and parent by checking both left/right possibilities.
func (m *MMR) findFamily(pos uint64, height uint8) (sibling, parent uint64, isRightChild, found bool) {
	offset := uint64(1) << (height + 1)

	// Try as left child first: parent = pos + offset, sibling = pos + offset - 1
	leftParent := pos + offset
	rightSibling := leftParent - 1

	if rightSibling < m.size && rightSibling != pos {
		rightNode, err := m.store.Get(rightSibling)
		if err == nil && rightNode.Height == height {
			if leftParent < m.size {
				parentNode, err := m.store.Get(leftParent)
				if err == nil && parentNode.Height == height+1 {
					return rightSibling, leftParent, false, true
				}
			}
		}
	}

	// Try as right child: parent = pos + 1, sibling = pos + 1 - offset
	rightParent := pos + 1
	if offset <= pos+1 {
		leftSibling := rightParent - offset

		if leftSibling < m.size && leftSibling != pos {
			leftNode, err := m.store.Get(leftSibling)
			if err == nil && leftNode.Height == height {
				if rightParent < m.size {
					parentNode, err := m.store.Get(rightParent)
					if err == nil && parentNode.Height == height+1 {
						return leftSibling, rightParent, true, true
					}
				}
			}
		}
	}

	// No valid family found - we're at a peak
	return 0, 0, false, false
}

// InclusionProof contains all data needed to verify a leaf's inclusion in the MMR.
type InclusionProof struct {
	LeafIndex    uint64         // Index of the proven leaf
	LeafHash     [32]byte       // Hash of the leaf node
	MerklePath   []ProofElement // Path from leaf to its peak
	Peaks        [][32]byte     // All peak hashes for bagging
	PeakPosition int            // Position of the relevant peak in Peaks slice
	MMRSize      uint64         // Size of the MMR when proof was generated
	Root         [32]byte       // The root at time of proof generation
}

// ProofElement represents a single step in a Merkle proof.
type ProofElement struct {
	Hash   [32]byte // Sibling hash
	IsLeft bool     // True if sibling is on the left
}

// Verify checks that this proof is valid for the given leaf data.
func (p *InclusionProof) Verify(leafData []byte) error {
	// Recompute leaf hash
	expectedLeafHash := HashLeaf(leafData)
	if expectedLeafHash != p.LeafHash {
		return ErrHashMismatch
	}

	// Walk up the Merkle path to the peak
	currentHash := p.LeafHash
	for _, elem := range p.MerklePath {
		if elem.IsLeft {
			currentHash = HashInternal(elem.Hash, currentHash)
		} else {
			currentHash = HashInternal(currentHash, elem.Hash)
		}
	}

	// Verify we reached the correct peak
	if p.PeakPosition >= len(p.Peaks) {
		return ErrInvalidProof
	}
	if currentHash != p.Peaks[p.PeakPosition] {
		return ErrInvalidProof
	}

	// Verify peak bagging produces the root
	if len(p.Peaks) == 1 {
		if p.Peaks[0] != p.Root {
			return ErrInvalidProof
		}
		return nil
	}

	// Bag peaks from right to left
	root := p.Peaks[len(p.Peaks)-1]
	for i := len(p.Peaks) - 2; i >= 0; i-- {
		root = HashInternal(p.Peaks[i], root)
	}

	if root != p.Root {
		return ErrInvalidProof
	}

	return nil
}

// ---------- MMR Geometry Functions ----------

// findPeaks returns the indices of all peak nodes for an MMR of the given size.
func findPeaks(size uint64) []uint64 {
	if size == 0 {
		return nil
	}

	var peaks []uint64
	pos := uint64(0)

	// Process each "mountain" from left to right
	for pos < size {
		// Find the largest perfect tree that fits
		height := highestPeak(size - pos)
		if height == 0 {
			peaks = append(peaks, pos)
			pos++
			continue
		}

		// Size of a perfect tree of this height
		treeSize := (uint64(1) << (height + 1)) - 1
		if pos+treeSize > size {
			// Tree doesn't fit, try smaller
			height--
			treeSize = (uint64(1) << (height + 1)) - 1
		}

		// Peak is at the rightmost position of this tree
		peakPos := pos + treeSize - 1
		peaks = append(peaks, peakPos)
		pos += treeSize
	}

	return peaks
}

// highestPeak returns the height of the largest perfect binary tree
// that can fit in the given size.
func highestPeak(size uint64) uint8 {
	if size == 0 {
		return 0
	}
	// The largest tree with 2^(h+1)-1 nodes that fits
	h := uint8(bits.Len64(size+1) - 1)
	for h > 0 {
		treeSize := (uint64(1) << (h + 1)) - 1
		if treeSize <= size {
			return h
		}
		h--
	}
	return 0
}

// leafCountFromSize converts MMR size (total nodes) to leaf count.
func leafCountFromSize(size uint64) uint64 {
	if size == 0 {
		return 0
	}

	// Count leaves by summing leaves in each mountain
	count := uint64(0)
	pos := uint64(0)

	for pos < size {
		height := highestPeak(size - pos)
		treeSize := (uint64(1) << (height + 1)) - 1
		if pos+treeSize > size {
			height--
			treeSize = (uint64(1) << (height + 1)) - 1
		}
		if treeSize == 0 {
			treeSize = 1
		}

		// Leaves in a perfect tree of height h is 2^h
		leaves := uint64(1) << height
		count += leaves
		pos += treeSize
	}

	return count
}

// findPeakContaining finds the peak index that contains the given leaf index.
func findPeakContaining(leafIndex, size uint64) uint64 {
	peaks := findPeaks(size)
	pos := uint64(0)

	for _, peak := range peaks {
		// Tree size from current position to peak
		treeSize := peak - pos + 1
		if leafIndex >= pos && leafIndex < pos+treeSize {
			return peak
		}
		pos += treeSize
	}

	// Default to last peak
	if len(peaks) > 0 {
		return peaks[len(peaks)-1]
	}
	return 0
}

// GetNodeHeight calculates the height of a node given its index in the MMR.
// Uses the property that a node at position p has height h if
// the binary representation of (p+1) has h trailing ones.
func GetNodeHeight(index uint64) uint8 {
	pos := index + 1
	height := uint8(0)

	// Count trailing ones in binary representation
	for pos&1 == 1 {
		height++
		pos >>= 1
	}

	// Adjust for MMR structure
	if height > 0 {
		height--
	}

	return height
}

File: rust/witnessd-core/src/mmr/mmr.rs
use crate::mmr::errors::MmrError;
use crate::mmr::node::Node;
use crate::mmr::proof::{InclusionProof, ProofElement, RangeProof};
use crate::mmr::store::Store;
use std::sync::RwLock;

pub struct MMR {
    store: Box<dyn Store>,
    state: RwLock<MmrState>,
}

struct MmrState {
    size: u64,
    peaks: Vec<u64>,
}

impl MMR {
    pub fn new(store: Box<dyn Store>) -> Result<Self, MmrError> {
        let size = store.size()?;
        let peaks = if size == 0 {
            Vec::new()
        } else {
            find_peaks(size)
        };
        Ok(Self {
            store,
            state: RwLock::new(MmrState { size, peaks }),
        })
    }

    pub fn append(&self, data: &[u8]) -> Result<u64, MmrError> {
        let mut state = self.state.write().unwrap();
        let leaf_index = state.size;
        let leaf = Node::new_leaf(leaf_index, data);
        self.store.append(&leaf)?;
        state.size += 1;

        loop {
            let peaks = find_peaks(state.size);
            if peaks.len() < 2 {
                state.peaks = peaks;
                break;
            }
            let last_idx = peaks[peaks.len() - 1];
            let prev_idx = peaks[peaks.len() - 2];
            let last = self.store.get(last_idx)?;
            let prev = self.store.get(prev_idx)?;
            if last.height != prev.height {
                state.peaks = peaks;
                break;
            }
            let new_node = Node::new_internal(state.size, last.height + 1, &prev, &last);
            self.store.append(&new_node)?;
            state.size += 1;
        }

        Ok(leaf_index)
    }

    pub fn get_peaks(&self) -> Result<Vec<[u8; 32]>, MmrError> {
        let state = self.state.read().unwrap();
        if state.size == 0 {
            return Ok(Vec::new());
        }
        let peaks = find_peaks(state.size);
        let mut hashes = Vec::with_capacity(peaks.len());
        for idx in peaks {
            hashes.push(self.store.get(idx)?.hash);
        }
        Ok(hashes)
    }

    pub fn get_root(&self) -> Result<[u8; 32], MmrError> {
        let peaks = self.get_peaks()?;
        if peaks.is_empty() {
            return Err(MmrError::Empty);
        }
        if peaks.len() == 1 {
            return Ok(peaks[0]);
        }
        let mut root = peaks[peaks.len() - 1];
        for i in (0..peaks.len() - 1).rev() {
            root = crate::mmr::node::hash_internal(peaks[i], root);
        }
        Ok(root)
    }

    pub fn size(&self) -> u64 {
        self.state.read().unwrap().size
    }

    pub fn leaf_count(&self) -> u64 {
        leaf_count_from_size(self.state.read().unwrap().size)
    }

    pub fn get(&self, index: u64) -> Result<Node, MmrError> {
        let state = self.state.read().unwrap();
        if index >= state.size {
            return Err(MmrError::IndexOutOfRange);
        }
        self.store.get(index)
    }

    pub fn get_leaf_index(&self, leaf_ordinal: u64) -> Result<u64, MmrError> {
        let state = self.state.read().unwrap();
        if state.size == 0 {
            return Err(MmrError::Empty);
        }
        let leaf_count = leaf_count_from_size(state.size);
        if leaf_ordinal >= leaf_count {
            return Err(MmrError::IndexOutOfRange);
        }
        let mut current_leaf = 0u64;
        for idx in 0..state.size {
            let node = self.store.get(idx)?;
            if node.height == 0 {
                if current_leaf == leaf_ordinal {
                    return Ok(idx);
                }
                current_leaf += 1;
            }
        }
        Err(MmrError::IndexOutOfRange)
    }

    pub fn get_leaf_indices(&self, start: u64, end: u64) -> Result<Vec<u64>, MmrError> {
        let state = self.state.read().unwrap();
        if start > end {
            return Err(MmrError::InvalidProof);
        }
        let leaf_count = leaf_count_from_size(state.size);
        if end >= leaf_count {
            return Err(MmrError::IndexOutOfRange);
        }
        let mut indices = Vec::with_capacity((end - start + 1) as usize);
        let mut current_leaf = 0u64;
        for idx in 0..state.size {
            let node = self.store.get(idx)?;
            if node.height == 0 {
                if current_leaf >= start && current_leaf <= end {
                    indices.push(idx);
                }
                current_leaf += 1;
                if current_leaf > end {
                    break;
                }
            }
        }
        Ok(indices)
    }

    pub fn generate_proof(&self, leaf_index: u64) -> Result<InclusionProof, MmrError> {
        let state = self.state.read().unwrap();
        if state.size == 0 {
            return Err(MmrError::Empty);
        }
        if leaf_index >= state.size {
            return Err(MmrError::IndexOutOfRange);
        }
        let node = self.store.get(leaf_index)?;
        if node.height != 0 {
            return Err(MmrError::InvalidProof);
        }
        let (path, peak_index) = self.generate_merkle_path(leaf_index)?;
        let peaks = self.get_peaks()?;
        let peak_indices = find_peaks(state.size);
        let mut peak_position = None;
        for (i, idx) in peak_indices.iter().enumerate() {
            if *idx == peak_index {
                peak_position = Some(i);
                break;
            }
        }
        let peak_position = peak_position.ok_or(MmrError::InvalidProof)?;
        let root = self.get_root()?;
        Ok(InclusionProof {
            leaf_index,
            leaf_hash: node.hash,
            merkle_path: path,
            peaks,
            peak_position,
            mmr_size: state.size,
            root,
        })
    }

    pub fn generate_range_proof(
        &self,
        start_leaf: u64,
        end_leaf: u64,
    ) -> Result<RangeProof, MmrError> {
        let state = self.state.read().unwrap();
        if state.size == 0 {
            return Err(MmrError::Empty);
        }
        if start_leaf > end_leaf {
            return Err(MmrError::InvalidProof);
        }
        let leaf_count = leaf_count_from_size(state.size);
        if end_leaf >= leaf_count {
            return Err(MmrError::IndexOutOfRange);
        }
        let leaf_indices = self.get_leaf_indices(start_leaf, end_leaf)?;
        let mut leaf_hashes = Vec::with_capacity(leaf_indices.len());
        for idx in &leaf_indices {
            leaf_hashes.push(self.store.get(*idx)?.hash);
        }
        let (sibling_path, peak_index) = self.generate_range_merkle_path(&leaf_indices)?;
        let peaks = self.get_peaks()?;
        let peak_indices = find_peaks(state.size);
        let mut peak_position = None;
        for (i, idx) in peak_indices.iter().enumerate() {
            if *idx == peak_index {
                peak_position = Some(i);
                break;
            }
        }
        let peak_position = peak_position.ok_or(MmrError::InvalidProof)?;
        let root = self.get_root()?;
        Ok(RangeProof {
            start_leaf,
            end_leaf,
            leaf_indices,
            leaf_hashes,
            sibling_path,
            peaks,
            peak_position,
            mmr_size: state.size,
            root,
        })
    }

    fn generate_merkle_path(&self, leaf_index: u64) -> Result<(Vec<ProofElement>, u64), MmrError> {
        let mut path = Vec::new();
        let mut pos = leaf_index;
        let node = self.store.get(pos)?;
        let mut height = node.height;

        loop {
            let (sibling_pos, parent_pos, is_right_child, found) = self.find_family(pos, height)?;
            if !found {
                return Ok((path, pos));
            }
            let sibling = self.store.get(sibling_pos)?;
            path.push(ProofElement {
                hash: sibling.hash,
                is_left: is_right_child,
            });
            pos = parent_pos;
            height += 1;
        }
    }

    fn find_family(&self, pos: u64, height: u8) -> Result<(u64, u64, bool, bool), MmrError> {
        let state = self.state.read().unwrap();
        let offset = 1u64 << (height + 1);

        let left_parent = pos + offset;
        let right_sibling = left_parent.saturating_sub(1);
        if right_sibling < state.size && right_sibling != pos {
            let right_node = self.store.get(right_sibling)?;
            if right_node.height == height {
                if left_parent < state.size {
                    let parent = self.store.get(left_parent)?;
                    if parent.height == height + 1 {
                        return Ok((right_sibling, left_parent, false, true));
                    }
                }
            }
        }

        let right_parent = pos + 1;
        if offset <= pos + 1 {
            let left_sibling = right_parent - offset;
            if left_sibling < state.size && left_sibling != pos {
                let left_node = self.store.get(left_sibling)?;
                if left_node.height == height {
                    if right_parent < state.size {
                        let parent = self.store.get(right_parent)?;
                        if parent.height == height + 1 {
                            return Ok((left_sibling, right_parent, true, true));
                        }
                    }
                }
            }
        }

        Ok((0, 0, false, false))
    }

    fn generate_range_merkle_path(
        &self,
        leaf_indices: &[u64],
    ) -> Result<(Vec<ProofElement>, u64), MmrError> {
        use std::collections::HashMap;
        if leaf_indices.is_empty() {
            return Err(MmrError::InvalidProof);
        }
        let mut covered: HashMap<u64, bool> = HashMap::new();
        for idx in leaf_indices {
            covered.insert(*idx, true);
        }
        let mut path: Vec<ProofElement> = Vec::new();
        let mut current_level: Vec<u64> = leaf_indices.to_vec();
        let mut height: u8 = 0;
        let mut peak_index = 0u64;

        while !current_level.is_empty() {
            current_level.sort_unstable();
            let mut next_level = Vec::new();
            let mut processed_parents: HashMap<u64, bool> = HashMap::new();
            for pos in &current_level {
                let (sibling_pos, parent_pos, is_right_child, found) =
                    self.find_family(*pos, height)?;
                if !found {
                    peak_index = *pos;
                    continue;
                }
                if *processed_parents.get(&parent_pos).unwrap_or(&false) {
                    continue;
                }
                processed_parents.insert(parent_pos, true);
                if !covered.get(&sibling_pos).copied().unwrap_or(false) {
                    let sibling = self.store.get(sibling_pos)?;
                    path.push(ProofElement {
                        hash: sibling.hash,
                        is_left: is_right_child,
                    });
                }
                covered.insert(parent_pos, true);
                next_level.push(parent_pos);
            }
            current_level = next_level;
            height += 1;
        }

        Ok((path, peak_index))
    }
}

pub fn find_peaks(size: u64) -> Vec<u64> {
    if size == 0 {
        return Vec::new();
    }
    let mut peaks = Vec::new();
    let mut pos = 0u64;
    while pos < size {
        let mut height = highest_peak(size - pos);
        if height == 0 {
            peaks.push(pos);
            pos += 1;
            continue;
        }
        let mut tree_size = (1u64 << (height + 1)) - 1;
        if pos + tree_size > size {
            height -= 1;
            tree_size = (1u64 << (height + 1)) - 1;
        }
        let peak_pos = pos + tree_size - 1;
        peaks.push(peak_pos);
        pos += tree_size;
    }
    peaks
}

pub fn highest_peak(size: u64) -> u8 {
    if size == 0 {
        return 0;
    }
    let bits_len = 64 - (size + 1).leading_zeros() as u8;
    let mut h = bits_len.saturating_sub(1);
    while h > 0 {
        let tree_size = (1u64 << (h + 1)) - 1;
        if tree_size <= size {
            return h;
        }
        h -= 1;
    }
    0
}

pub fn leaf_count_from_size(size: u64) -> u64 {
    if size == 0 {
        return 0;
    }
    let mut count = 0u64;
    let mut pos = 0u64;
    while pos < size {
        let mut height = highest_peak(size - pos);
        let mut tree_size = (1u64 << (height + 1)) - 1;
        if pos + tree_size > size {
            height -= 1;
            tree_size = (1u64 << (height + 1)) - 1;
        }
        if tree_size == 0 {
            tree_size = 1;
        }
        count += 1u64 << height;
        pos += tree_size;
    }
    count
}
